executor/new_executor_xapi.go

```diff
@@ -229,69 +226,79 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
-		e.buildTableTasks(handles)
-		limitCount := int64(-1)
-		if e.indexPlan.LimitCount != nil {
-			limitCount = *e.indexPlan.LimitCount
-		}
-		concurrency := 2
-		if limitCount == -1 {
-			concurrency = len(e.tasks)
-		} else if limitCount > int64(BaseLookupTableTaskSize) {
-			concurrency = int(limitCount/int64(BaseLookupTableTaskSize) + 1)
-		}
-		log.Debugf("[TIME_INDEX_TABLE_CONCURRENT_SCAN] start %d workers", concurrency)
-		e.runTableTasks(concurrency)
+		idxResult.Fetch()
+
+		// use a background goroutine to fetch index, put the result in e.tasks.
+		// e.tasks serves as a pipeline, so fetch index and get table data would
+		// run concurrency.
+		e.tasks = make(chan *lookupTableTask, 10)
+		go e.fetchHandles(idxResult, e.tasks)
 	}
+
 	for {
-		if e.taskCursor >= len(e.tasks) {
-			return nil, nil
-		}
-		task := e.tasks[e.taskCursor]
-		if !task.done {
-			startTs := time.Now()
-			err := <-task.doneCh
-			if err != nil {
-				return nil, errors.Trace(err)
+		if e.taskCurr == nil {
+			taskCurr, ok := <-e.tasks
+			if !ok {
+				return nil, e.tasksErr
 			}
-			task.done = true
-			log.Debugf("[TIME_INDEX_TABLE_SCAN] time: %v handles: %d", time.Now().Sub(startTs), len(task.handles))
+			e.taskCurr = taskCurr
 		}
-		if task.cursor < len(task.rows) {
-			row := task.rows[task.cursor]
-			task.cursor++
-			return row, nil
+
+		row, err := e.taskCurr.getRow()
+		if err != nil || row != nil {
+			return row, errors.Trace(err)
 		}
-		e.taskCursor++
+		e.taskCurr = nil
 	}
 }
 
-// fetchHandle fetches all handles from the index.
-// This should be optimized to fetch handles in batch.
-func (e *NewXSelectIndexExec) fetchHandles() ([]int64, error) {
-	idxResult, err := e.doIndexRequest()
-	if err != nil {
-		return nil, errors.Trace(err)
+func (e *NewXSelectIndexExec) fetchHandles(idxResult xapi.SelectResult, ch chan<- *lookupTableTask) {
+	defer close(ch)
+
+	workCh := make(chan *lookupTableTask)
+	defer close(workCh)
+	limitCount := -1
+	if e.indexPlan.LimitCount != nil {
+		limitCount = int(*e.indexPlan.LimitCount)
 	}
-	idxResult.Fetch()
-	handles, err := extractHandlesFromIndexResult(idxResult)
-	if err != nil {
-		return nil, errors.Trace(err)
+	concurrency := 2
+	for i := 0; i < concurrency; i++ {
+		go e.pickAndExecTask(workCh)
 	}
-	if !e.indexPlan.OutOfOrder {
-		// Save the index order.
-		e.indexOrder = make(map[int64]int)
-		for i, h := range handles {
-			e.indexOrder[h] = i
+
+	for {
+		handles, err := extractHandlesFromIndexResult(idxResult)
+		if err != nil || handles == nil {
+			e.tasksErr = errors.Trace(err)
+			return
+		}
+
+		if !e.indexPlan.OutOfOrder {
+			// Save the index order.
+			e.indexOrder = make(map[int64]int)
+			for i, h := range handles {
+				e.indexOrder[h] = i
+			}
+		}
+
+		tasks := e.buildTableTasks(handles)
+		if limitCount == -1 {
+			limitCount = len(tasks)
+		}
+		for concurrency < limitCount {
+			go e.pickAndExecTask(workCh)
```

Comment:
of course not.
`workCh` is a channel, one data in it (`*lookupTableTask`) can't be dequeue twice.